metapool/prep.py

```diff
@@ -938,3 +907,178 @@ def demux_pre_prep(pre_prep):
+                         ', '.join(sorted(not_fastqs)))
+
+    for sample_file in sample_files:
+        # make no assumptions about the naming format of fastq files, other
+        # than their names will begin with a valid sample_id. This is to
+        # support non-Illumina generated fastq files.
+        _, file_name = split(sample_file)
+
+        # for each sample_file, find the subset of sample_ids that match the
+        # beginning of the file's name and assume that the longest matching
+        # sample_id is the best and proper match. Ex: myfile_sre.fastq.gz
+        # should match both the sample_ids [myfile_sre, myfile] but we want to
+        # ensure myfile_sre is the correct match, and myfile_sre.fastq.gz is
+        # not associated with myfile and myfile.fastq.gz is not associated with
+        # myfile_sre.fastq.gz.
+        matches = [s_id for s_id in sample_ids if file_name.startswith(s_id)]
+
+        # sort the matches so the longest matching sample_id will be first.
+        matches = sorted(matches, key=len, reverse=True)
+
+        if not matches:
+            # if no sample_ids could be matched to this file, then something
+            # is wrong.
+            raise ValueError(f"{sample_file} could not be matched to any "
+                             "sample-id")
+
+        # map sample_ids to a list of matching sample_files. We should expect
+        # two matches for R1 and R2 files.
+        results[matches[0]].append(sample_file)
+
+    unmatched_sample_ids = set(sample_ids) - set(list(results.keys()))
+
+    # after all possible sample_files have been matched to sample_ids, ensure
+    # that the number of matching files for each sample_id is 2 and only 2,
+    # _assuming_ that one is an R1 file and the other is an R2 file.
+    msgs = []
+    for sample_id in results:
+        count = len(results[sample_id])
+        if count != 2:
+            msg = sorted(results[sample_id])
+            msgs.append(f"{sample_id} matched an unexpected number of samples "
+                        f"({msg})")
+    if msgs:
+        raise ValueError("\n".join(msgs))
+
+    return results, unmatched_sample_ids
+
+
+def _find_filtered_files(fp):
+    # assume that only fastq.gz files are desired and it's okay to ignore
+    # other odd files that we find.
+
+    # assume that all desired fastq.gz files will be stored in a directory
+    # named 'filtered_sequences'. This will automatically ignore files in
+    # 'only-adapter-filtered', 'zerofiles' and possibly other directories
+    # in the future as well.
+
+    # we expect the filepaths to be of the form:
+    # <project_name_w_qiita_id>/filtered_sequences/<fastq_file>
+    # However, we intentionally search across all directories to locate
+    # any fastq files w/in a filtered_sequences subdirectory, even if
+    # they don't match our expectation so that no potentially desirable
+    # file is silently filtered out.
```

Comment:
Is this actually possible? I'm just wondering when/how would this happen!?

Reply:
In this code, paths are passed around, split, munged, and rejoined often. In my experience paths are fairly brittle in that a broken assumption or quick change in code can cause a function like this to mistakenly miss some or all of the intended files and/or include other ones like the adapter-trimmed files.

The idea here is to never let a fastq file be silently ignored. Take only files found in filtered_sequences directories but make it known if a filtered_sequences directory full of files was placed either deeper or shallower than expected due to an error.

The value in this approach is that we're still just using a glob() call with a ** wildcard. We're not wasting resources traversing extra directories when they don't exist. If for some reason they do exist we won't ignore them.